### ISSUE TITLE
Conv2D question

### ISSUE THREAD

**Issue by krworks** [Dobiasd/frugally-deep#70]
I've noticed a bit of issues with the matrix gemm calls from conv2d.  My first inclination to improve performance on conv2d was to enable OpenMP support in eigen.  I defined what google defined in their tensorflow usage of eigen (define EIGEN_USE_THREADS) and compiled -O3 with openmp support in gcc.

After verifying that it is certainly using the OpenMP code, I saw the results were slower (2-4x) on any input I gave it, large and small. (1,32,24 ... 1,240,320, etc).  I tried various thread counts from 2 to 4 to 8 to 16.  They're all terrible.  Also Eigen is adding over 1MB of bloat to the executable and I think it's too complex a library for my needs.

I'm in a place now where I would like to research how to use the gemm reference functor:
seen here (https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/gemm_functors.h).  I realize this is for reference only but I would love to see how I could integrate it in so I can get my hands on optimizing the mat ops myself.  I've gotten major, major gains using parallelism and simd on other layers (2x or 4x on BatchNorm, etc) but this particular one is more complicated of course.

I would like to use the following with frugally-deep's conv2d im2col ([A] matrix) and filter matrix [B].  It does seem that frugally-deep does create im2col mat/filter mat rather differently from google's implementation here:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/conv_ops_using_gemm.cc

The reference function to me is very beautiful and I would like to see how I can approach proper parallelism by optimizing inputs to a fork of this:
```
template <class T1, class T2, class T3>
class ReferenceGemmFunctor {
 public:
  void operator()(tensorflow::OpKernelContext* ctx, size_t m, size_t n,
                  size_t k, const T1* a, size_t lda, const T2* b, size_t ldb,
                  T3* c, size_t ldc) {
    const size_t a_i_stride = lda;
    const size_t a_l_stride = 1;
    const size_t b_j_stride = 1;
    const size_t b_l_stride = ldb;
    const size_t c_i_stride = ldc;
    const size_t c_j_stride = 1;
    size_t i, j, l;
    for (j = 0; j < n; j++) {
      for (i = 0; i < m; i++) {
        T3 total(0);
        for (l = 0; l < k; l++) {
          const size_t a_index = ((i * a_i_stride) + (l * a_l_stride));
          const T1 a_value = a[a_index];
          const size_t b_index = ((j * b_j_stride) + (l * b_l_stride));
          const T2 b_value = b[b_index];
          total += (a_value * b_value);
        }
        const size_t c_index = ((i * c_i_stride) + (j * c_j_stride));
        c[c_index] = total;
      }
    }
  }
};
```

If you are not interested allowing swapping of different gemm functors in such as this, could you let me know how it could be done? I seem to be at a loss with the differences between frugally-deep im2col and tensorflow's.  It seems to go beyond just channel first/last but rather theirs is created multiple times with massive amounts of zeros filled in at each parallel step. (Something that I think is quite weird).

One more note: I believe that even single CPU mode in Tensor flow will not disable parallelism inside of Eigen.  So gains between tensorflow / frugally-deep has a lot to do with major flaws in eigen's OpenMP.

**Comment by krworks** [comment]
I've managed to get the reference gemm working with the im2col and filter mat inside of conv2d.  It's slower by about 5x but thats without any SIMD or alignment optimizations.  I'm not sure if alignment is totally possible yet but SIMD definitely is.

The call is: (a = im2col, b = filter_mat, c = output)
    gemm_reference(
        b_rows, a_cols, b_cols,
        b, b_cols,
        a, a_cols,
        c, c_cols); 

It's good to actually see the math happening with my own eyes.  Sorry for being verbose and possibly off topic but OpenMP is most certainly worth benchmarking over a wide variety of inputs vs disabled MP.  I honestly cannot get it working faster at all under any circumstances so I really don't know what's going on there.
